### PR TITLE
file picker should let you select CDF JSON

### PR DESCRIPTION
**Pull request by tarheel**
bitmoji

### CODE REVIEW

File: src/main/java/com/rcv/GuiConfigController.java [hunk 1]
```diff
@@ -317,19 +318,20 @@ public void buttonAddCvrFileClicked() {
     String cvrFilePath = textFieldCvrFilePath.getText().trim();
     String cvrFirstVoteCol = textFieldCvrFirstVoteCol.getText().trim();
     String cvrFirstVoteRow = textFieldCvrFirstVoteRow.getText().trim();
+    String cvrProvider = textFieldCvrProvider.getText().trim();
     if (cvrFilePath.isEmpty()) {
       Logger.log(Level.WARNING, "CVR file path is required!");
-    } else if (cvrFirstVoteCol.isEmpty()) {
+    } else if (cvrFirstVoteCol.isEmpty() && !cvrProvider.equals("CDF")) {
```
Comment (HEdingfield) [inline src/main/java/com/rcv/GuiConfigController.java]: This seems like a really bizarre UX to me to have to count on the user to type "CDF" exactly in the provider box in order to skip needing to enter row and column fields, especially considering provider is an optional field. Why not instead just check if the file type is .json?

File: src/main/resources/GuiConfigLayout.fxml [hunk 2]
```diff
@@ -103,9 +103,9 @@
                     text="Select"/>
                 </HBox>
                 <HBox alignment="CENTER_LEFT" spacing="4.0">
-                  <TextField prefHeight="25.0" prefWidth="160.0" promptText="First Vote Col Index *"
+                  <TextField prefHeight="25.0" prefWidth="160.0" promptText="First Vote Col Index"
```
Comment (HEdingfield) [inline src/main/resources/GuiConfigLayout.fxml]: Please update the descriptions in the txt file if you're going to change the "required" behavior here.

**Comment by HEdingfield** [comment]
I forgot to add that the precinct test is broken here too, as is test_set_0 (which I assume is expected since the other PR fixes that). However, it's odd to me that test_set_1 and beyond pass here but fail in #260. Probably just because I'm out of the loop...

File: src/main/java/com/rcv/GuiConfigController.java (hunk 1 above)
Comment (tarheel) [reply src/main/java/com/rcv/GuiConfigController.java]: OK, that makes sense here.

**Comment by tarheel** [comment]
> I forgot to add that the precinct test is broken here too, as is test_set_0 (which I assume is expected since the other PR fixes that). However, it's odd to me that test_set_1 and beyond pass here but fail in #260. Probably just because I'm out of the loop...

Don't worry! I'll make sure everything is working when I'm done with the other PR.

File: src/main/java/com/rcv/GuiConfigController.java [hunk 3]
```diff
@@ -317,19 +318,21 @@ public void buttonAddCvrFileClicked() {
     String cvrFilePath = textFieldCvrFilePath.getText().trim();
     String cvrFirstVoteCol = textFieldCvrFirstVoteCol.getText().trim();
     String cvrFirstVoteRow = textFieldCvrFirstVoteRow.getText().trim();
+    String cvrProvider = textFieldCvrProvider.getText().trim();
```
Comment (HEdingfield) [inline src/main/java/com/rcv/GuiConfigController.java]: You can inline this again in line 335.

File: src/main/resources/config_file_documentation.txt [hunk 4]
```diff
@@ -58,12 +58,12 @@ Config file must be valid JSON format. Examples can be found under the test fold
         example: /Users/test_data/2015-portland-mayor-cvr.xlsx
         value: string of length [1..1000]
 
-      "firstVoteColumnIndex" required
+      "firstVoteColumnIndex" required for Excel files
```
Comment (HEdingfield) [inline src/main/resources/config_file_documentation.txt]: "for Excel files" seems a little vague. Maybe "required unless cast-vote records are in JSON format"? This applies to all instances where "Excel" is used here.

File: src/main/resources/config_file_documentation.txt [hunk 5]
```diff
@@ -87,7 +87,7 @@ Config file must be valid JSON format. Examples can be found under the test fold
         if not supplied: none
 
   "candidates" required
-    list of registered candidate names and associated candidate code
+    list of registered candidate names and associated candidate code (note: leave empty when CVR is in Common Data Format)
```
Comment (HEdingfield) [inline src/main/resources/config_file_documentation.txt]: Probably get rid of the * in "Name *" in `GuiConfigLayout.fxml` in light of this. Though at this point I'm feeling like these notes are getting pretty watered down. Maybe we should consider a new symbol (**?) for "required unless CVR is JSON"?

Another thing: you may need to modify the config validation to allow for no candidates if JSON.

Comment (tarheel) [reply src/main/resources/config_file_documentation.txt]: Well, "Name" is still required if you're actually going to add a candidate. We just don't want you to add any if the input is CDF. I'll enforce that in a separate PR.

> Another thing: you may need to modify the config validation to allow for no candidates if JSON.

This part already works.

**Pull request merged**